File: Julia/textbook_track/chapter01/ba1c.jl
# Find the Reverse Complement of a String

const NUCLEOTIDE_COMPLEMENTS = Dict('A' => 'T', 'C' => 'G', 'T' => 'A', 'G' => 'C')


calc_reverse_complement(genome) = join(map(n -> NUCLEOTIDE_COMPLEMENTS[n], Iterators.reverse(genome)), "")


function main()
    dna = readline()
    result = calc_reverse_complement(dna)
    println(result)
end


main()

File: OCaml/src/textbook_track/chapter01/ba1c.ml
open Batteries

let nucleotide_complements = [('A', 'T'); ('C', 'G'); ('T', 'A'); ('G', 'C')]


let calc_reverse_complement (dna : string) : string =
    String.(dna |> map (fun nucleotide -> List.assoc nucleotide nucleotide_complements) |> rev)


let () =
    let dna = read_line () in
    let result = calc_reverse_complement dna in
    print_endline result
